### PR TITLE
feat(ecs): Alarm-based Rollbacks adding new datastructure,Constructor, enableDeploymentAlarms and MetricDeploymentAlarms

### LINKED ISSUE #24344: (aws-ecs): Provide support for DeploymentAlarms on ECS Service
### Describe the feature

The ECS team recently introduced CloudWatch alarms for ECS deployments as described in https://aws.amazon.com/about-aws/whats-new/2022/12/amazon-ecs-cloudwatch-alarms-safety-deployments/

It would be great if ECS Service and ECS Fargate service could support this via constructor props 

Fargate docs:
* https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-ecs.FargateService.html
* https://docs.aws.amazon.com/cdk/api/v1/docs/aws-ecs-readme.html

### Use Case

This feature allows customers to add automated safeguards for Amazon ECS service rolling updates. You can now monitor and automatically react to changes during an Amazon ECS rolling update by using Amazon CloudWatch alarms. This enables you to more easily automate discovery and remediation for failed deployments and minimize the impact of a bad change.

### Proposed Solution

Provide a props argument or property that is able to populate the `DeploymentAlarms` attribute on an ECS Service as per https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-service-deploymentalarms.htmlcan

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

v2.66.1

### Environment details (OS name and version, etc.)

MacOS

### PR DESCRIPTION
**Pull request by homakk**
Adding a new interface, deploymentAlarmConfig, to allow customers to use existing alarms when the services are constructed. 
Enable Deployment Alarms which take advantage of arbitrary alarms and configure them after service initialization.
Add an alarm based on ECS metric which appended to the list of deployment alarms at synthesis time
   *   Customize the alarm threshold and evaluationPeriods using 2nd argument ecsMetricAlarmProps.
   *   `threshold  value 85`
   *   `evaluationPeriods value 3`
   
   Reference documentation : https://aws.amazon.com/blogs/containers/automate-rollbacks-for-amazon-ecs-rolling-deployments-with-cloudwatch-alarms/ 
   
image

>
> Describe the reason for this change, what the solution is, and any
> important design decisions you made. 
>
> Remember to follow the [CONTRIBUTING GUIDE] and [DESIGN GUIDELINES] for any
> code you submit.
>
> [CONTRIBUTING GUIDE]: https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md
> [DESIGN GUIDELINES]: https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts
Comment (bvtujo) [inline packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts:1287]: This config isn't valid, you can't use EXTERNAL deployment controller with deployment alarm config https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-service-deploymentalarms.html

File: packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts
Comment (bvtujo) [inline packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts:1296]: why are we testing for this? It's not relevant to the deployment alarms config. Please take this out of the rest of the tests and make sure that the deployment controller is valid.

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:3]: Hooo so we can use `cloudwatch.IAlarm` , gotcha

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts]: modified!

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts]: when we try to use positive instead of negative it giving an error that 

Whether to enable the deployment alarms. If this property is defined, alarms will be implicitly enabled.
@default
- disabled
This comparison appears to be unintentional because the types 'AlarmBehavior.FAIL_ON_ALARM' and 'AlarmBehavior.ROLLBACK_ON_ALARM' have no overlap.

So I'm rewriting this as ` rollback: props.deploymentAlarms.behavior !== AlarmBehavior.FAIL_ON_ALARM,`

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts]: Gotcha, at present Im not changing anything, I will change this in future when I working on Convenience methods. Thanks for letting me know about this in advance.

File: packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts:1287]: Gotcha, 
So we can use ECS deployment controller with deploymemt alarm config.

File: packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts]: Sure.
I removed this from all the test cases.

File: packages/@aws-cdk/aws-ecs/README.md
Comment (pallymore) [inline packages/@aws-cdk/aws-ecs/README.md:626]: > ECS will conduct the deployment as normal. Then ...

I wonder if we missed a sentence before this one 😅

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (pallymore) [inline packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:101]: I'm curious if this can be an empty array - what happens if we set alarm config with `alarms: []` ?

Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:101]: This validation will be added in subsequent PR , so alarms should be minimum one.

File: packages/@aws-cdk/aws-ecs/README.md
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/README.md]: Ya I missed Given a list of alarms, before ECS will conduct .......
I added it.

**Comment by pallymore** [comment]
thanks for addressing the review comments 😬 let's make sure the codebuild check passes

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts
Comment (pallymore) [inline packages/@aws-cdk/aws-ecs/lib/cluster.ts:59]: let's remove these merge markers 

also on line 76

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts
Comment (pallymore) [inline packages/@aws-cdk/aws-ecs/lib/cluster.ts:73]: actually - should these be part of this PR? I don't see how this is related to the deployment alarms changes.

File: packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts
Comment (pallymore) [inline packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts:1296]: i'm curious why those `as` are needed - especially there is another one below. is there a way to avoid using `as` ?

**Review by pallymore — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/cluster.ts:59]: Thank you so much for guiding, I removed those stashed changes, I don't know why these commits appear in my PR when I cherry-pick the latest commit which is related to alarmrollback one

File: packages/@aws-cdk/aws-ecs/lib/cluster.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/cluster.ts:73]: Sir that's true, but when I did push the PR directly after working on my perious task which is related to defaultCapacityproviders, there is so much clumsy in PR 

So I did following steps, To keep the extraneous commits to a minimum in this and other PRs. It looks like there’s only one commit that has to do with the data structures and constructor method, so the PR branch should only contain that commit and any future ones I make.

1.pull down the latest main branch from aws-cdk origin
2.rename your current branch to something else (abr-backup or similar)
3.cherry-pick that one commit onto a new branch named “alarambasedrollback” (original name of this current branch)
4.force push the new branch to your fork of the cdk to get rid of the extra stuff


When I did that that commit is also coming to this PR, I don't know why but it appears.

File: packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts
Comment (bvtujo) [reply packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts:1296]: I think we should use the `Template.fromStack()` matcher pattern that's found elsewhere in the test files: 

```typescript
const template = Template.fromStack(stack);
template.hasResourceProperties('AWS::ECS::Service', {
  DeploymentConfiguration: {
    Alarms: {
      ... // Your desired configuration here
    },
  },
});
```

Comment (bvtujo) [reply packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts:1296]: You can use the CFN spec to find the right field names: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-service-deploymentconfiguration.html

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (bvtujo) [inline packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:72]: Let's make this more descriptive: 
```suggestion
 * Deployment behavior when an ECS Service Deployment Alarm is triggered.
```

**Review by bvtujo — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts]: ya did it

File: packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts]: Ya we can use that as austin said, Thank you austin
 const template = Template.fromStack(stack);
      template.hasResourceProperties('AWS::ECS::Service', {
        DeploymentConfiguration: {
          Alarms: {
            Enable: true,
            Rollback: true,
            AlarmNames: [myAlarm.alarmName],
          },
        },
      });

Comment (homakk) [reply packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts]: I updated the test cases by using Template.fromStack() matcher pattern

File: packages/@aws-cdk/aws-ecs/test/ec2/integ.deployment-alarms.ts
Comment (bvtujo) [inline packages/@aws-cdk/aws-ecs/test/ec2/integ.deployment-alarms.ts:51]: Let's be sure to change this integ test stack name to something that fits the test better. `DeploymentAlarms` or something similar.

**Review by bvtujo — COMMENTED** [review]
LGTM. My only remaining comment is that we should probably not merge this into main yet, but instead use this branch as a feature branch. You should base your next PRs on this branch so that we can merge all the changes for alarm based rollbacks into main in one fell swoop and we don't run the risk of partially releasing this feature.

File: packages/@aws-cdk/aws-ecs/test/ec2/integ.deployment-alarms.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/test/ec2/integ.deployment-alarms.ts]: Oops sorry, It my mistake I forgot to change that. I updated that with latest snapshots.

File: packages/@aws-cdk/aws-ecs/test/ec2/integ.deployment-alarms.ts
Comment (bvtujo) [inline packages/@aws-cdk/aws-ecs/test/ec2/integ.deployment-alarms.ts:44]: You don't need to manually specify this; the default deployment controller is `ECS`. But we should make sure to add validation logic and test for it in `ec2-service.test.ts` where we error out if [deployment controller](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-service-deploymentcontroller.html) is "EXTERNAL" or "CODE_DEPLOY". 

This is true of `DeploymentCircuitBreaker` too (can't be enabled when DeploymentController is not ECS), so you can piggyback on any existing validation to also check DeploymentAlarms.

**Review by bvtujo — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-ecs/test/ec2/integ.deployment-alarms.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/test/ec2/integ.deployment-alarms.ts]: Gotcha,
I added following validation with new test case for this. 
`if (props.deploymentAlarms
        && deploymentController
        && deploymentController.type !== DeploymentControllerType.ECS) {
      Annotations.of(this).addError('Deployment alarms requires the ECS deployment controller.');
    }`
    
   And
   also updated the integration file with latest snapshots

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:619]: Ok 
so to validate the final state of the resource we can directly throw an error in the validation for deployment alarms,
As Annotations used in the circuit breaker I thought we need to use that in the deployment alarms,
Now I got the point 
So instead of using  `Annotations.of(this).addError` we can use `throw new Error ('Deployment circuit breaker requires the ECS deployment controller.');`

thanks for the clarification.

**Review by bvtujo — APPROVED** [review]

File: packages/@aws-cdk/aws-ecs/README.md
Comment (iamhopaul123) [inline packages/@aws-cdk/aws-ecs/README.md:626]:
```suggestion
(https://aws.amazon.com/blogs/containers/automate-rollbacks-for-amazon-ecs-rolling-deployments-with-cloudwatch-alarms/) Given a list of alarms, ECS will conduct the deployment as normal. Then, it will monitor the alarms for a “bake time” determined by the alarm’s sampling and thresholds. If any alarm enters the “ALARM” state during the bake time, the deployment will either fail (requiring roll-forward) or be rolled back. 
```

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (iamhopaul123) [inline packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:74]:
```suggestion
 * Units of this metrics are percentages
```

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (iamhopaul123) [inline packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:738]: Are we using this anywhere?

**Review by iamhopaul123 — COMMENTED** [review]
Can you help me verify if it works for "removing alarm-based rollback config"? Let's say I've already have ABR configured, and now if I want to stop using it, can I just remove the config in CDK and deploy again?

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:738]: Yes we are using this for Customize the alarm threshold and evaluationPeriods using 2nd argument ecsMetricAlarmProps and I also added validations in it .
I added unit test cases for the validations present in it.

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:74]: Got it!I 
I made the sentence correction.

File: packages/@aws-cdk/aws-ecs/README.md
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/README.md:626]: changes made in the latest commit.

**Comment by homakk** [comment]
> Can you help me verify if it works for "removing alarm-based rollback config"? Let's say I've already have ABR configured, and now if I want to stop using it, can I just remove the config in CDK and deploy again?

For this I'm not completely aware , I think if you already have ABR configured try to remove the config in cdk and deploy again!

**Comment by iamhopaul123** [comment]
> For this I'm not completely aware , I think if you already have ABR configured try to remove the config in cdk and deploy again!

That's what I meant...For some reason (update service ignore fields if they are not specified, unless you explicitly disable them) having config removed in CDK/CFN can't really disassociate the alarms. For example, we have this [fix](https://github.com/aws/copilot-cli/pull/4565) in Copilot

**Comment by homakk** [comment]
> > For this I'm not completely aware , I think if you already have ABR configured try to remove the config in cdk and deploy again!
> 
> That's what I meant...For some reason (update service ignore fields if they are not specified, unless you explicitly disable them) having config removed in CDK/CFN can't really disassociate the alarms. For example, we have this [fix](https://github.com/aws/copilot-cli/pull/4565) in Copilot

But here we just add the alarm , not removing the alarm , if alarm is empty then it will throw the error
```ts // Throw an error if deployment alarms are already configured
    if (this.deploymentAlarms) {
      throw new Error('Deployment alarms are already configured.');
    }
    // Throw an error if alarms array is empty
    if (alarmConfig.alarms.length === 0) {
      throw new Error('Specify at least one deployment alarm');
    }```

**Comment by iamhopaul123** [comment]
Yes but then how do users use CDK to disassociate alarms that were associated with the service before?

**Comment by homakk** [comment]
> disassociate

Ok but this was not mentioned in the documention https://quip-amazon.com/4UCiAqFFKS1I/Alarm-based-Rollbacks-for-CDK which I have. 
so do I need to add that in this PR or can we do it later ?

**Comment by iamhopaul123** [comment]
I would expect disassociating is there if we add support for associating ABR 🤔 Otherwise it's surprising to see it is possible to associate ABR to an ECS service but no way to disassociate the alarms.

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (iamhopaul123) [inline packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:738]: I think when users try to disassociate, we can just set `deployAlarms` to an empty string list, and `enable` to be false as is now.

Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:738]: gotcha
so I made alarmNames: [ ] 
keeping enable : false as it is

**Review by iamhopaul123 — APPROVED** [review]
LGTM feel free to merge the PR after some manual testing to make sure some common scenarios work!

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (pallymore) [inline packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:757]: I'm not familiar with this part but the code below doesn't seem to be service connect specific. 

the members in `EcsMetric` can be used for regular services though 🤔  right?

Comment (iamhopaul123) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:757]: Sorry I missed this. +1 to Yurui. ABR should not depend on service connect.

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (iamhopaul123) [inline packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:740]: Why do we need this?

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (iamhopaul123) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:753]: I tried to search for `createEcsMetricAlarm` and can't find any reference. Can you paste a link where it get used?

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (pallymore) [inline packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:767]: Hi - I might be missing the context here - looks like if I set `useAsDeploymentAlarm` to `false` - this code would still create a default alarm for me? (also it's using values from `ecsMetricAlarmProps.alarmProps` here

could you explain what's the use case for `useAsDeploymentAlarm === false`

Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts]: I made changes as follows so that if we set  useAsDeploymentAlarm === false it wont create deployment alarms for us.
Please check the latest commit for updated changes

```
if (ecsMetricAlarmProps?.useAsDeploymentAlarm) {
      if (!this.deploymentAlarms) {
        this.deploymentAlarms = {
          enable: true,
          rollback: true,
          alarmNames: [],
        };
      }
      this.deploymentAlarms.alarmNames.push(metricAlarm.node.id);
```

**Review by pallymore — APPROVED** [review]
LGTM 🎊

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (iamhopaul123) [inline packages/@aws-cdk/aws-ecs/lib/base/base-service.ts:836]: why do we throw this error instead of letting them to override?

**Review by iamhopaul123 — COMMENTED** [review]
In addition to specify ABR alarms in literal when creating the ECS service, do we want to have another method like `service.addRollbackAlarms(...)` or similar like that for import experience ?

**Comment by homakk** [comment]
> The pull request linter has failed. See the aws-cdk-automation comment below for failure reasons. If you believe this pull request should receive an exemption, please comment and provide a justification.
> 
> A comment requesting an exemption should contain the text `Exemption Request`. Additionally, if clarification is needed add `Clarification Request` to a comment.

Exemption Request

File: packages/@aws-cdk/aws-ecs/lib/base/base-service.ts
Comment (homakk) [reply packages/@aws-cdk/aws-ecs/lib/base/base-service.ts]: Yeah, that true we can do that.
I just removed that error, so it can override

**Review by iamhopaul123 — APPROVED** [review]

**Pull request closed without merging**
